File: tests/helpers/db_client.py
from __future__ import annotations

import os
from collections.abc import Iterator
from contextlib import contextmanager
from typing import Any

import psycopg
from psycopg.rows import dict_row


def _database_url() -> str:
    raw = os.getenv(
        "DATABASE_URL_SYNC",
        os.getenv(
            "DATABASE_URL",
            "postgresql://securedox:securedox_local_pw@localhost:5432/securedox",
        ),
    )
    return raw.replace("postgresql+psycopg://", "postgresql://").replace(
        "postgresql+asyncpg://", "postgresql://"
    )


@contextmanager
def db_connection() -> Iterator[psycopg.Connection[dict[str, Any]]]:
    with psycopg.connect(_database_url(), row_factory=dict_row) as connection:
        yield connection


def get_document(document_id: str) -> dict[str, Any] | None:
    with db_connection() as connection:
        with connection.cursor() as cursor:
            cursor.execute(
                """
                select id::text, tenant_id, document_type::text, status::text,
                       original_filename, correlation_id, processed_at
                  from documents
                 where id = %s
                """,
                (document_id,),
            )
            return cursor.fetchone()


def audit_events_for_document(document_id: str) -> list[dict[str, Any]]:
    with db_connection() as connection:
        with connection.cursor() as cursor:
            cursor.execute(
                """
                select id::text, tenant_id, document_id::text, action::text,
                       actor, correlation_id, detail, created_at
                  from audit_events
                 where document_id = %s
                 order by created_at asc
                """,
                (document_id,),
            )
            return list(cursor.fetchall())

File: packages/test-framework/src/clients/DbClient.ts
import pg from "pg";
import type { Pool as PoolType } from "pg";

const { Pool } = pg;

export type AuditEventRow = {
  id: string;
  tenant_id: string;
  document_id: string | null;
  action: string;
  actor: string;
  correlation_id: string;
  detail: Record<string, unknown>;
  created_at: Date;
};

export type DocumentRow = {
  id: string;
  tenant_id: string;
  status: string;
  document_type: string;
  original_filename: string;
  correlation_id: string;
  processed_at: Date | null;
};

export class DbClient {
  private readonly pool: PoolType;

  constructor(
    connectionString =
      process.env.DATABASE_URL_SYNC ??
      process.env.DATABASE_URL?.replace("postgresql+asyncpg://", "postgresql://") ??
      "postgresql://securedox:securedox_local_pw@localhost:5432/securedox"
  ) {
    this.pool = new Pool({
      connectionString: connectionString.replace("postgresql+psycopg://", "postgresql://")
    });
  }

  async close(): Promise<void> {
    await this.pool.end();
  }

  async getDocument(documentId: string): Promise<DocumentRow | null> {
    const result = await this.pool.query<DocumentRow>(
      `select id::text, tenant_id, status::text, document_type::text, original_filename,
              correlation_id, processed_at
         from documents
        where id = $1`,
      [documentId]
    );
    return result.rows[0] ?? null;
  }

  async auditEventsForDocument(documentId: string): Promise<AuditEventRow[]> {
    const result = await this.pool.query<AuditEventRow>(
      `select id::text, tenant_id, document_id::text, action::text, actor,
              correlation_id, detail, created_at
         from audit_events
        where document_id = $1
        order by created_at asc`,
      [documentId]
    );
    return result.rows;
  }

  async auditEventsByCorrelation(correlationId: string): Promise<AuditEventRow[]> {
    const result = await this.pool.query<AuditEventRow>(
      `select id::text, tenant_id, document_id::text, action::text, actor,
              correlation_id, detail, created_at
         from audit_events
        where correlation_id = $1
        order by created_at asc`,
      [correlationId]
    );
    return result.rows;
  }
}
